### PR TITLE
Writing of PNG images.

### PR DESCRIPTION
**Pull request by local-minimum**
This writes both raw pixels and compressed jpeg as png in the same output directory as the csvs using the pattern `nanoseconds.png`.

* Fixes so the RID/SID of continuation and stop packages are discarded in the aggregator
* Uses the PacketLength in the innosat reader.

### CODE REVIEW

File: internal/aez/ccd_data.go [hunk 1]
```diff
@@ -114,6 +114,12 @@ func (gain CCDGain) Truncation() uint8 {
 	return uint8(gain & 0b1111)
 }
 
+// JPEGQUncompressed16bit is the value for non-12bit image data
+var JPEGQUncompressed16bit = uint8(101)
```
Comment (skymandr) [inline internal/aez/ccd_data.go]: why not `const`?

File: internal/exports/disk_writer.go [hunk 2]
```diff
@@ -61,9 +57,51 @@ func DiskCallbackFactory(
...
+					shift,
+					imgFileName,
+				)
+
+				imgFile, err := os.Create(imgFileName)
+				if err != nil {
+					log.Printf("failed creating %s: %s", imgFileName, err)
+					panic(err.Error())
+				}
+				defer imgFile.Close()
```
Comment (skymandr) [inline internal/exports/disk_writer.go:102]: move to line 98?

Comment (skymandr) [inline internal/exports/disk_writer.go]:
```suggestion
				imgFile, err := os.Create(imgFileName)
				defer imgFile.Close()
				if err != nil {
					log.Printf("failed creating %s: %s", imgFileName, err)
					panic(err.Error())
				}

```
Or is that bad, i.e. will `Close` in itself likely cause an error?

File: internal/extractors/innosat_reader.go [hunk 3]
```diff
@@ -16,6 +16,9 @@ type SourcePackage struct {
 	ApplicationPayload []byte
 }
 
+var crcChecksumLength int = 2
```
Comment (skymandr) [inline internal/extractors/innosat_reader.go]: Again, why not `const`? (I'm not sure what's best practice here.)

File: internal/aez/ccd_data.go (hunk 1 above)
Comment (local-minimum) [inline internal/aez/ccd_data.go]:
```suggestion
const JPEGQUncompressed16bit = uint8(101)

```

Comment (local-minimum) [reply internal/aez/ccd_data.go]: I forgot about them :shrug:

File: internal/common/exportable.go [hunk 4]
```diff
@@ -14,4 +14,6 @@ type ExportablePackage interface {
 	CSVSpecifications() []string
 	OriginName() string
 	AEZData() interface{}
+	RemainingBuffer() []byte
```
Comment (joakimmoller) [inline internal/common/exportable.go:17]: This is a general comment on the interfaces:
1. I think the interfaces should be named after what they do, not what "data" it holds (be cause interfaces does not contain data it is just set of methods)
1. Exportable is a subset of Exportable package - could be dragged in...

```
type ExportablePackage interface {
    Exportable
	OriginName() string
    .
```

Better names could be somethinge like CVSExporter and CVSPackageExporter?

File: internal/extractors/innosat_reader.go (hunk 3 above)
Comment (local-minimum) [inline internal/extractors/innosat_reader.go]:
```suggestion
const crcChecksumLength int = 2
```

File: internal/exports/disk_writer.go (hunk 2 above)
Comment (local-minimum) [reply internal/exports/disk_writer.go]: no that should be fine

File: internal/aez/ccd_data.go [hunk 5]
```diff
@@ -114,6 +114,13 @@ func (gain CCDGain) Truncation() uint8 {
 	return uint8(gain & 0b1111)
 }
 
+// JPEGQUncompressed16bit is the value for non-12bit image data
+const JPEGQUncompressed16bit = uint8(101)
+
+
+// NCOLStartOffset says how many more columns than reported the actual columns are
+var NCOLStartOffset uint16 = 1
```
Comment (skymandr) [inline internal/aez/ccd_data.go]:
```suggestion
const NCOLStartOffset uint16 = 1

```

File: internal/extractors/innosat_reader.go [hunk 6]
```diff
@@ -16,6 +16,9 @@ type SourcePackage struct {
 	ApplicationPayload []byte
 }
 
+const crcChecksumLength int = 2
+var pusLengthOffset int = 1
```
Comment (skymandr) [inline internal/extractors/innosat_reader.go]:
```suggestion
const pusLengthOffset int = 1
```

File: internal/extractors/aggregator.go [hunk 7]
```diff
@@ -16,13 +16,13 @@ func Aggregator(target chan<- common.DataRecord, source <-chan common.DataRecord
 	multiPackBuffer := bytes.NewBuffer([]byte{})
 	var multiPackStarted bool
 	var multiPackStart common.DataRecord
+	var sidRidLength = 2
```
Comment (skymandr) [inline internal/extractors/aggregator.go]:
```suggestion
	const sidRidLength = 2
```
maybe?

File: internal/exports/images.go [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+			filename,
+			nPixels,
+			width,
+			height,
+			width*height,
+		)
+	}
+	buf := bytes.NewBuffer([]byte{})
+	if shift > 0 {
+		for idx := 0; idx < nPixels; idx++ {
```
Comment (joakimmoller) [inline internal/exports/images.go]: mutation of your own object then assigning back... 
```
for idx, pix := range pixels {
    pixels[idx] = pix << shift
}
```

File: internal/extractors/aggregator.go (hunk 7 above)
Comment (joakimmoller) [reply internal/extractors/aggregator.go]: or binary.sizeof(uint16)

File: internal/extractors/innosat_reader.go [hunk 9]
```diff
@@ -35,9 +41,11 @@ func DecodeSource(data []byte) (SourcePackage, error) {
 		return SourcePackage{}, err
 	}
 
+	sliceStart := binary.Size(header) + binary.Size(tmPayload)
+	sliceEnd := sliceStart + int(header.PacketLength) - binary.Size(tmPayload) - crcChecksumLength + pusLengthOffset
```
Comment (joakimmoller) [inline internal/extractors/innosat_reader.go:45]: This was not the error. The following line is better inline with the spec.
`sliceEnd := binary.Size(header) + int(header.PacketLength) - crcChecksumLength + pusLengthOffset`
even though
`sliceEnd := len(data) - crcChecksumLength `
still works.

File: internal/common/datarecord_test.go [hunk 10]
```diff
@@ -359,3 +359,87 @@ func TestDataRecord_OriginName(t *testing.T) {
 		})
 	}
 }
+
+func TestDataRecord_RemainingBuffer(t *testing.T) {
```
Comment (joakimmoller) [inline internal/common/datarecord_test.go:363]: I think you should skip this test, you are adding 40 lines of boilerplate code for me to read for testing a getter of its simplest form.

File: internal/common/datarecord_test.go [hunk 11]
```diff
@@ -359,3 +359,87 @@ func TestDataRecord_OriginName(t *testing.T) {
...
+				Buffer:       tt.fields.Buffer,
+			}
+			if got := record.RemainingBuffer(); !reflect.DeepEqual(got, tt.want) {
+				t.Errorf("DataRecord.RemainingBuffer() = %v, want %v", got, tt.want)
+			}
+		})
+	}
+}
+
+func TestDataRecord_ParsingError(t *testing.T) {
```
Comment (joakimmoller) [inline internal/common/datarecord_test.go:406]: This test is for a direct getter - skip this.

**Comment by joakimmoller** [comment]
In general nice and clean. I have some minor suggestions...

**Review by joakimmoller — APPROVED** [review]

File: internal/common/exportable.go (hunk 4 above)
Comment (local-minimum) [reply internal/common/exportable.go:17]: I appended it to #48 
When creating the interface I wasn't sure it it was going to be only for the csv exporting or also other formats but now that it seems to only need to cover our csv-needs `CSVExporter` might be better than `Exporter` (you can add that to the issue if you think so).

File: internal/common/datarecord_test.go (hunk 10 above)
Comment (local-minimum) [reply internal/common/datarecord_test.go:363]: I'll leave it now to have time to merge this, and with the #48 , the getter would be removed anyhow.

File: internal/exports/images.go [hunk 12]
```diff
@@ -0,0 +1,99 @@
...
+			pixels[idx] = pixels[idx] << shift
```
Comment (local-minimum) [inline internal/exports/images.go]:
```suggestion
		for idx, pix := range pixels {
			pixels[idx] = pix << shift

```

File: internal/extractors/innosat_reader.go (hunk 9 above)
Comment (local-minimum) [reply internal/extractors/innosat_reader.go:45]: Not sure if you want me to change anything?

**Pull request merged**